exporters/memory/include/opentelemetry/exporters/memory/in_memory_metric_exporter.h

```diff
@@ -0,0 +1,79 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <atomic>
+#include <vector>
+
+#include "opentelemetry/exporters/memory/in_memory_data.h"
+#include "opentelemetry/sdk/metrics/export/metric_producer.h"
+#include "opentelemetry/sdk/metrics/instruments.h"
+#include "opentelemetry/sdk/metrics/push_metric_exporter.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace exporter
+{
+namespace memory
+{
+
+using InMemoryMetricData = InMemoryData<sdk::metrics::ResourceMetrics>;
+
+/// A Push Metric Exporter which accumulates metrics data in memory and allows it to be inspected.
+/// It is not thread-safe.
+class InMemoryMetricExporter final : public sdk::metrics::PushMetricExporter
+{
+  using AggregationTemporalityMap =
+      std::map<sdk::metrics::InstrumentType, sdk::metrics::AggregationTemporality>;
+
+public:
+  /// @param buffer_size a required value that sets the size of the CircularBuffer
+  /// @param temporality Output temporality as a function of instrument kind.
+  InMemoryMetricExporter(size_t buffer_size,
+                         const sdk::metrics::AggregationTemporalitySelector &temporality)
+      : data_(std::make_shared<InMemoryMetricData>(buffer_size)), temporality_(temporality)
+  {}
+
+  InMemoryMetricExporter(const InMemoryMetricExporter &)  = delete;
+  InMemoryMetricExporter(const InMemoryMetricExporter &&) = delete;
+  void operator=(const InMemoryMetricExporter &)          = delete;
+  void operator=(const InMemoryMetricExporter &&)         = delete;
+  ~InMemoryMetricExporter() override                      = default;
+
+  sdk::common::ExportResult Export(const sdk::metrics::ResourceMetrics &data) noexcept override;
+  sdk::metrics::AggregationTemporality GetAggregationTemporality(
+      sdk::metrics::InstrumentType instrument_type) const noexcept override;
+  bool ForceFlush(
+      std::chrono::microseconds timeout = (std::chrono::microseconds::max)()) noexcept override;
+  bool Shutdown(std::chrono::microseconds timeout = std::chrono::microseconds(0)) noexcept override;
+
+  const std::shared_ptr<InMemoryMetricData> &GetData() const;
+
+private:
+  std::shared_ptr<InMemoryMetricData> data_;
+  std::atomic<bool> is_shutdown_{false};
+  sdk::metrics::AggregationTemporalitySelector temporality_;
+};
+
+/// A factory for InMemoryMetricExporter
+class InMemoryMetricExporterFactory
```

Comment:
Please declare the factory in a separate header file.

This is not just style, it has some far reaching consequences in the SDK ABI: an application should be able to invoke the factory create, and get an instance back, without having to know the exact binary layout of an in memory exporter. Putting both in the same header file defeats this.

Reply:
Done.